File: leetcode_python/Two_Pointers/backspace-string-compare.py
"""

844. Backspace String Compare
Easy

Given two strings s and t, return true if they are equal when both are typed into empty text editors. '#' means a backspace character.

Note that after backspacing an empty text, the text will continue empty.

 

Example 1:

Input: s = "ab#c", t = "ad#c"
Output: true
Explanation: Both s and t become "ac".
Example 2:

Input: s = "ab##", t = "c#d#"
Output: true
Explanation: Both s and t become "".
Example 3:

Input: s = "a#c", t = "b"
Output: false
Explanation: s becomes "c" while t becomes "b".
 

Constraints:

1 <= s.length, t.length <= 200
s and t only contain lowercase letters and '#' characters.
 

Follow up: Can you solve it in O(n) time and O(1) space?

"""

# V0
class Solution(object):
    def backspaceCompare(self, S, T):

        def check(x):
            ans = []
            for i in x:
                if i != "#":
                    ans.append(i)
                # both `if ans:` and `if len(ans) > 0` work
                #if ans:
                if len(ans) > 0:
                    if ans[-1] != "#" and i == "#":
                        ans.pop()
            return ans 
            
        return check(S) == check(T)

# V1
# http://bookshadow.com/weblog/2018/06/03/leetcode-backspace-string-compare/
class Solution(object):
    def backspaceCompare(self, S, T):
        """
        :type S: str
        :type T: str
        :rtype: bool
        """
        def toString(S):
            ans = []
            for c in S:
                if c == '#': ans and ans.pop()
                else: ans.append(c)
            return ''.join(ans)
        return toString(S) == toString(T)

# V1'
# http://bookshadow.com/weblog/2018/06/03/leetcode-backspace-string-compare/
class Solution(object):
    def backspaceCompare(self, S, T):
        def toString(S):
            ans = []
            #print ('S = ', S)
            for c in S:
                #print ('c = ', c )
                #print ('ans = ', ans )
                
                if c == '#': 
                    ans and ans.pop()
                else: 
                    ans.append(c)
                #print ('ans = ', ans )
            return ''.join(ans)
        ## re-run the runc compare m treated S, T 
        return toString(S) == toString(T)

# V1''
# https://www.jiuzhang.com/solution/backspace-string-compare/#tag-highlight-lang-python
class Solution(object):
    def backspaceCompare(self, S, T):
        def build(S):
            ans = []
            for c in S:
                if c != '#':
                    ans.append(c)
                elif ans:
                    ans.pop()
            return "".join(ans)
        return build(S) == build(T)

# V2
import itertools
class Solution(object):
    def backspaceCompare(self, S, T):
        """
        :type S: str
        :type T: str
        :rtype: bool
        """
        def findNextChar(S):
            skip = 0
            for i in reversed(range(len(S))):
                if S[i] == '#':
                    skip += 1
                elif skip:
                    skip -= 1
                else:
                    yield S[i]
        return all(x == y for x, y in itertools.zip_longest(findNextChar(S), findNextChar(T)))

File: leetcode_java/src/main/java/LeetCodeJava/TwoPointer/BackspaceStringCompare.java
package LeetCodeJava.TwoPointer;

// https://leetcode.com/problems/backspace-string-compare/
/**
 *

 844. Backspace String Compare
 Solved
 Easy
 Topics
 premium lock icon
 Companies
 Given two strings s and t, return true if they are equal when both are typed into empty text editors. '#' means a backspace character.

 Note that after backspacing an empty text, the text will continue empty.



 Example 1:

 Input: s = "ab#c", t = "ad#c"
 Output: true
 Explanation: Both s and t become "ac".
 Example 2:

 Input: s = "ab##", t = "c#d#"
 Output: true
 Explanation: Both s and t become "".
 Example 3:

 Input: s = "a#c", t = "b"
 Output: false
 Explanation: s becomes "c" while t becomes "b".


 Constraints:

 1 <= s.length, t.length <= 200
 s and t only contain lowercase letters and '#' characters.


 Follow up: Can you solve it in O(n) time and O(1) space?
 *
 *
 *
 */
import java.util.Stack;

public class BackspaceStringCompare {

    // V0
    // IDEA: STACK + SB
    public boolean backspaceCompare(String s, String t) {
        // edge

        Stack<String> st1 = new Stack<>();
        Stack<String> st2 = new Stack<>();

        for (String x : s.split("")) {
            if (x.equals("#")) {
                if (!st1.isEmpty()) {
                    st1.pop();
                }
            } else {
                st1.push(x);
            }
        }

        for (String x : t.split("")) {
            if (x.equals("#")) {
                if (!st2.isEmpty()) {
                    st2.pop();
                }
            } else {
                st2.push(x);
            }
        }

        StringBuilder sb1 = new StringBuilder();
        StringBuilder sb2 = new StringBuilder();

        while (!st1.isEmpty()) {
            sb1.append(st1.pop());
        }

        while (!st2.isEmpty()) {
            sb2.append(st2.pop());
        }

//        System.out.println(">>> sb1 = " + sb1.toString());
//        System.out.println(">>> sb2 = " + sb2.toString());

        return sb1.toString().contentEquals(sb2);
    }

    // V0-1
    // IDEA : STACK + stringBuilder
    /**
     * time = O(N)
     * space = O(1)
     */
    public boolean backspaceCompare_0_1(String s, String t) {

        if (s == null && t == null){
            return true;
        }

        if (s == null || t == null){
            return false;
        }

        String s_ = backSpaceStr(s);
        String t_ = backSpaceStr(t);
        return s_.equals(t_);
    }

    private String backSpaceStr(String input){
        Stack<String> stack = new Stack<>();
        for(String x: input.split("")){
            if (!x.equals("#")){
                stack.add(x);
            }else{
                if(!stack.isEmpty()){
                    stack.pop();
                }
            }
        }
        // stack to str
        StringBuilder sb = new StringBuilder();
        while(!stack.isEmpty()){
            sb.append(stack.pop());
        }

        return sb.toString();
    }


    // V1
    /**
     * time = O(N)
     * space = O(1)
     */
    public boolean backspaceCompare_1(String s, String t) {

        if (s == null && t == null){
            return true;
        }

        // TODO : check why can't have this condition
//        if ( (s == null || t != null) || (s != null && t == null) ){
//            return false;
//        }

        String _s = backSpaceOp(s);
        String _t = backSpaceOp(t);

//
//        System.out.println("_s = " + _s);
//        System.out.println("_t = " + _t);

        if (_s == null && _t == null){
            return true;
        }

        return _s.equals(_t);
    }


    private static String backSpaceOp(String s){

        if (s == null){
            return "";
        }

        Stack<String> stack = new Stack<>();

        for (int i = 0; i < s.length(); i++){
            String cur = String.valueOf(s.charAt(i));
            System.out.println("cur = " + cur + ", stack = " + stack.toString() + " is # ?" + cur.equals("#"));
            if (!cur.equals("#")){
                stack.push(cur);
            }else{
                if (!stack.empty()){
                    stack.pop();
                }
            }
        }

        String res = "";
        while (!stack.isEmpty()){
            String _cur = stack.pop();
            res += _cur;
        }

        return res;
    }

    // V2
    // IDEA : BUILD STRING
    // https://leetcode.com/problems/backspace-string-compare/editorial/
    /**
     * time = O(N)
     * space = O(1)
     */
    public boolean backspaceCompare_2(String S, String T) {
        return build(S).equals(build(T));
    }

    /**
     * time = O(N)
     * space = O(1)
     */
    public String build(String S) {
        Stack<Character> ans = new Stack();
        for (char c: S.toCharArray()) {
            if (c != '#')
                ans.push(c);
            else if (!ans.empty())
                ans.pop();
        }
        return String.valueOf(ans);
    }



    
}
